feat: upgrade mongodb-erlang library to support 5.1+

kjellwinblad opened the pull request:
This subsumes a similar PR in main-v4.4-enterprise https://github.com/emqx/emqx-enterprise/pull/1692. I did not realize that the community edition also has a dependency on MongoDB.

@lafirest @zmstone , We need a Chinese translation for the changelog entry.

thalesmg commented on changes/v4.4.15-en.md:
```diff
@@ -0,0 +1,7 @@
+# v4.4.15
+
+## Enhancements
+- The MongoDB library has been upgraded to support MongoDB version 5.1 and greater.
```
```suggestion
- The MongoDB library has been upgraded to support MongoDB version 5.1 and greater [#9707](https://github.com/emqx/emqx/pull/9707).
```

thalesmg commented on changes/v4.4.15-en.md:
```diff
@@ -0,0 +1,7 @@
+# v4.4.15
+
+## Enhancements
+- The MongoDB library has been upgraded to support MongoDB version 5.1 and greater [#9707](https://github.com/emqx/emqx/pull/9707).
```
Reminder: need to add ZH translation (a first approximation may be obtained with [deepl](https://www.deepl.com/translator)).

kjellwinblad replied:
Very cool. I added a translation from deepl. At least it made sense according to one Chinese non-technical person that I asked.

kjellwinblad commented:
@thalesmg thanks for the review!

kjellwinblad commented:
The "Check App Version" test fails. See [check_apps_version_4_x (24.3.4.2-1, ubuntu20.04)](https://github.com/emqx/emqx/actions/runs/3875988609/jobs/6609273780#logs). I have not managed to spot what is wrong. There seems to be a downgrade entry from 3.0.19 to 3.0.18 and the appup file seems to be parseable. I would appreciate if someone else can have a look as well (@thalesmg).

thalesmg commented:
> The "Check App Version" test fails. See [check_apps_version_4_x (24.3.4.2-1, ubuntu20.04)](https://github.com/emqx/emqx/actions/runs/3875988609/jobs/6609273780#logs). I have not managed to spot what is wrong. There seems to be a downgrade entry from 3.0.19 to 3.0.17 with the pattern "3.0.[0-9]+". I would appreciate if someone else can have a look as well to see if a mistake can be spoted (@thalesmg).

The check is reporting that the following instructions are missing from this version:

```erlang
#{down =>
                          [{<<"3.0.18">>,
                            [{load_module,mongoc,brutal_purge,soft_purge,[]},
                             {load_module,mongo_protocol,brutal_purge,soft_purge,[]},
                             {load_module,mc_worker_logic,brutal_purge,soft_purge,[]},
                             {load_module,mc_worker,brutal_purge,soft_purge,[]},
                             {load_module,mc_utils,brutal_purge,soft_purge,[]},
                             {load_module,mc_topology_logics,brutal_purge,soft_purge,[]},
                             {load_module,mc_topology,brutal_purge,soft_purge,[]},
                             {load_module,mc_super_sup,brutal_purge,soft_purge,[]},
                             {load_module,mc_monitor,brutal_purge,soft_purge,[]},
                             {load_module,mc_cursor,brutal_purge,soft_purge,[]},
                             {load_module,mc_connection_man,brutal_purge,soft_purge,[]},
                             {delete_module,mc_worker_pid_info}]}],
                      up =>
                          [{<<"3.0.18">>,
                            [{load_module,mongoc,brutal_purge,soft_purge,[]},
                             {load_module,mongo_protocol,brutal_purge,soft_purge,[]},
                             {load_module,mc_worker_logic,brutal_purge,soft_purge,[]},
                             {load_module,mc_worker,brutal_purge,soft_purge,[]},
                             {load_module,mc_utils,brutal_purge,soft_purge,[]},
                             {load_module,mc_topology_logics,brutal_purge,soft_purge,[]},
                             {load_module,mc_topology,brutal_purge,soft_purge,[]},
                             {load_module,mc_super_sup,brutal_purge,soft_purge,[]},
                             {load_module,mc_monitor,brutal_purge,soft_purge,[]},
                             {load_module,mc_cursor,brutal_purge,soft_purge,[]},
                             {load_module,mc_connection_man,brutal_purge,soft_purge,[]},
                             {add_module,mc_worker_pid_info}]}]}
```

It seems that for `3.0.18` there's only 1 instruction indeed:

https://github.com/emqx/mongodb-erlang/blob/c2bfa8f84377f939079e4f1a425d680e4da8bb00/src/mongodb.appup.src#L3-L5

thalesmg commented:
Aha, there's some overlap in the version specifications.  There's this explicit `3.0.18` entry, then later there's the regex you mentioned that also matches `3.0.18`.

kjellwinblad commented:
> Aha, there's some overlap in the version specifications. There's this explicit 3.0.18 entry, then later there's the regex you mentioned that also matches 3.0.18.

Yeah, my assumption was that the first matching one would be used but seems like that assumption is wrong...

kjellwinblad commented:
> The check is reporting that the following instructions are missing from this version: ...

> Aha, there's some overlap in the version specifications. There's this explicit 3.0.18 entry, then later there's the regex you mentioned that also matches 3.0.18.

I still do not understand what the problem is. With this version I assume you mean 3.0.19 which is the one that is specified in this PR. I don't think EMQX need to care about 3.0.18 as the previous version that EMQX used was 3.0.17. But even if 3.0.18 is relevant, only one module changed between 3.0.18 and 3.0.19 and that module is included in both of the matching instruction lists (<<"3.0.18">> and <<"3.0.[0-9]+">>).

I will try to make the versions non-overlapping. Maybe this is the cause of the error?

kjellwinblad commented:
Updated https://github.com/emqx/mongodb-erlang/ to make the patterns in the appup non-overlapping and replaced the tag v3.0.19 so that it points to the updated code.

kjellwinblad commented:
The script still report the same error even after the change. Seems to me like there is a problem with the script. First of all it should not need to be concerned about 3.0.18 at all as the last version included in EMQX was 3.0.17. Secondly, if I understand the error message correctly it reports that a lot of instructions should be added for downgrading (3.0.19 -> 3.0.18) and upgrading (3.0.18 -> 3.0.19) but only one module changed between 3.0.18 and 3.0.19.

thalesmg commented:
I think you must add these instructions reported by CI to [`3.0.18`](https://github.com/emqx/mongodb-erlang/blob/c0fe7976c17937df88f6f782e23e50e95ec877c9/src/mongodb.appup.src#L4):

```erlang
{load_module,mongoc,brutal_purge,soft_purge,[]},
{load_module,mongo_protocol,brutal_purge,soft_purge,[]},
{load_module,mc_worker_logic,brutal_purge,soft_purge,[]},
{load_module,mc_worker,brutal_purge,soft_purge,[]},
{load_module,mc_utils,brutal_purge,soft_purge,[]},
{load_module,mc_topology_logics,brutal_purge,soft_purge,[]},
{load_module,mc_topology,brutal_purge,soft_purge,[]},
{load_module,mc_super_sup,brutal_purge,soft_purge,[]},
{load_module,mc_monitor,brutal_purge,soft_purge,[]},
{load_module,mc_cursor,brutal_purge,soft_purge,[]},
{load_module,mc_connection_man,brutal_purge,soft_purge,[]},
{add_module,mc_worker_pid_info}]}
```

kjellwinblad commented:
For anyone interested in the solution to the appup problem discussed here. The check appup files scripts appropriately tries to skip versions of external dependencies that have not been used in an EMQX release. However, the script cannot do that when the version is specified as a binary in the appup file and not as a regular list string (binaries in appup files are interpreted as regexp patterns while list strings are interpreted as normal strings). So the solution was just to change the representation of 3.0.18 from <<"3.0.18">>  to "3.0.18". I have updated the emqx/mongodb-erlang repo so I think this is ready to be merged now but we need an approval from the code owner (@zmstone).
